windows/entry.cpp

```diff
@@ -7,19 +7,27 @@ struct uiEntry {
 	void (*onChanged)(uiEntry *, void *);
 	void *onChangedData;
 	BOOL inhibitChanged;
+	void (*onFinished)(uiEntry *, void *);
+	void *onFinishedData;
 };
 
 static BOOL onWM_COMMAND(uiControl *c, HWND hwnd, WORD code, LRESULT *lResult)
 {
 	uiEntry *e = uiEntry(c);
 
-	if (code != EN_CHANGE)
-		return FALSE;
-	if (e->inhibitChanged)
-		return FALSE;
-	(*(e->onChanged))(e, e->onChangedData);
-	*lResult = 0;
-	return TRUE;
+	if (code == EN_CHANGE) {
+    	if (e->inhibitChanged)
+	    	return FALSE;
+    	(*(e->onChanged))(e, e->onChangedData);
+	    *lResult = 0;
+    	return TRUE;
+    }
+	if (code == EN_KILLFOCUS) {
```

Comment:
Windows was the thing I was the least sure about what constitutes "finished editing", because the traditional wisdom was (as far as I can gather) to validate everything when dismissing a dialog box. My original idea for validating correctness of user entries was to provide a function `uiEntryAlertInvalid()` that would display a notification that didn't inhibit entry (on Windows, it'd use `EM_SHOWBALLOONTIP`) and you'd do this on the Changed event, but I'm not sure what people would prefer doing :/ (Pre-libui package ui actually did have this, but I've now forgotten what I called the method back then.)

That being said, I'd suggest keeping these in mind:

https://blogs.msdn.microsoft.com/oldnewthing/20040419-00/?p=39753
https://blogs.msdn.microsoft.com/oldnewthing/20050808-16/?p=34673
https://blogs.msdn.microsoft.com/oldnewthing/20090420-00/?p=18523

Reply:
Isn't dialog validation a slightly moot point anyway, without dialog box support?
I'd guess that a hypothetical dialog box would have some OnDismiss() callback that's invoked when OK or cancel is pressed. And that's where dialog-validation-as-a-whole code would go. I don't _think_ that handling `EN_KILLFOCUS` here would preclude hooking into whatever windows (and gtk and osx).

Caveat: I have no experience or knowledge whatsoever about dialog boxes on any of the target platforms :-)

The use case I'm interested in myself is less to do with validation, and more to do with reacting to a "completed" change, rather than the higher-frequency OnChanged() callbacks. In particular, I have a search box where the searches take a few seconds to perform, so I only want to perform them when the user has completed entering their search query.

Reply:
And yes, my reading of those Raymond Chen posts seems to suggest that `WM_KILLFOCUS` is the tricky one, rather than `EN_FOCUS`.